docs/discounts-lists/index.md

```diff
@@ -0,0 +1,496 @@
+  `Js.Array.mapi` it’s `(item, index)`, but for `List.mapi` it’s `(index,
+  item)`.
+- We have to add a call to
+  [Stdlib.Array.of_list](https://melange.re/v2.2.0/api/re/melange/Stdlib/Array/#val-of_list)
+  in between the calls to `List.mapi` and `React.array` to convert the list to
+  an array. Whenever we want to render a list of `React.element`s, we must first
+  convert it to an array of `React.element`s. To understand why, recall that a
+  list is just an object in the JS runtime, and React cannot directly render
+  objects.
+
+Because `Stdlib` is automatically opened, normally we can just call
+`Array.of_list`, but we have to use the full name `Stdlib.Array.of_list` because
+our custom `Array` module takes precedence[^2].
+
+To get all your code compiling again, you have to also refactor `Index`. But
+there all you have to do is change the array delimiters (`[||]`) to list
+delimiters (`[]`).
+
+## `List.nth_opt`
+
+If we peruse the `List` module a bit, we'll find a function that can make
+`Discount.getFreeBurger` shorter:
+[List.nth](https://melange.re/v1.0.0/api/re/melange/Stdlib/List/#val-nth). It
+takes an index `n` that returns the `n`-th element of a list. However, from
+previous experience, we don't want to use unsafe functions like this.
+Fortunately, there's a similar
+[List.nth_opt](https://melange.re/v1.0.0/api/re/melange/Stdlib/List/#val-nth_opt)
+function that does the same thing but is safer because it returns `option`
+instead of raising an exception. Let's refactor `Discount.getFreeBurger` to use
+it:
+
+```reason
+let getFreeBurger = (items: list(Item.t)) => {
+  items
+  |> List.filter(item =>
+       switch (item) {
+       | Item.Burger(_) => true
+       | Sandwich(_)
+       | Hotdog => false
+       }
+     )
+  |> List.map(Item.toPrice)
+  |> List.sort((x, y) => - compare(x, y))
+  |> List.nth_opt(1);
+};
+```
+
+By using `List.nth_opt`, we can simplify the function to a single expression.
+However, we get a compilation error:
+
+```text
+File "docs/order-confirmation/Discount.re", line 61, characters 18-19:
+61 |   |> List.nth_opt(1);
+                       ^
+Error: This expression has type int but an expression was expected of type
+         'a list
+```
+
+## Placeholder operator
+
+This is because of the type signature of `List.nth_opt` is
+
+```text
+list('a) => int => option('a)
+```
+
+That is, it accepts the list as the first argument, not the last. Recall that
+the pipe last operator (`|>`) pipes values into the last argument of a function.
+However, there's a way to override the placement of the argument:
+
+```reason
+|> List.nth_opt(_, 1)
+```
+
+Do not confuse the `_` here for wildcard, here it's a placeholder for where the
+argument should go. When we put `_` in the first argument position, it overrides
+the default behavior of the pipe last operator. Run `npm run test` to confirm
+that `Discount.getFreeBurger` works the same as before.
```

Comment:
I personally find this wildcard operator very hacky because it's only available in Reason syntax (so it's not possible to convert to OCaml). Maybe we could just get the result from the code above in a `prices` list? Like it had before. Then just leave `List.nth_opt(prices, 1)` at the end.

That way, we skip the compilation error, and the need for this whole section to explain some functionality that is not really "The OCaml Way" :)

Reply:
Yeah, it's not very OCaml Way, but I decided to be a little pragmatic and write about this because I anticipate talking about Belt in a later chapter. At the moment, using the placeholder operator is the only way to use most of the Belt API with the pipe last operator. I don't feel very strongly about it though, I could certainly just move this section into whatever future chapter actually brings up Belt. Wdyt?